pkg/proxy/nftables/proxier.go

```diff
@@ -0,0 +1,1631 @@
+)
+
+const sysctlNFConntrackTCPBeLiberal = "net/netfilter/nf_conntrack_tcp_be_liberal"
+
+// internal struct for string service information
+type servicePortInfo struct {
+	*proxy.BaseServicePortInfo
+	// The following fields are computed and stored for performance reasons.
+	nameString             string
+	clusterPolicyChainName string
+	localPolicyChainName   string
+	externalChainName      string
+}
+
+// returns a new proxy.ServicePort which abstracts a serviceInfo
+func newServiceInfo(port *v1.ServicePort, service *v1.Service, bsvcPortInfo *proxy.BaseServicePortInfo) proxy.ServicePort {
+	svcPort := &servicePortInfo{BaseServicePortInfo: bsvcPortInfo}
+
+	// Store the following for performance reasons.
+	svcName := types.NamespacedName{Namespace: service.Namespace, Name: service.Name}
+	svcPortName := proxy.ServicePortName{NamespacedName: svcName, Port: port.Name}
+	svcPort.nameString = svcPortName.String()
+
+	chainNameBase := servicePortChainNameBase(&svcPortName, strings.ToLower(string(svcPort.Protocol())))
+	svcPort.clusterPolicyChainName = servicePortPolicyClusterChainNamePrefix + chainNameBase
+	svcPort.localPolicyChainName = servicePortPolicyLocalChainNamePrefix + chainNameBase
+	svcPort.externalChainName = serviceExternalChainNamePrefix + chainNameBase
+
+	return svcPort
+}
+
+// internal struct for endpoints information
+type endpointInfo struct {
+	*proxy.BaseEndpointInfo
+
+	chainName       string
+	affinitySetName string
+}
+
+// returns a new proxy.Endpoint which abstracts a endpointInfo
+func newEndpointInfo(baseInfo *proxy.BaseEndpointInfo, svcPortName *proxy.ServicePortName) proxy.Endpoint {
+	chainNameBase := servicePortEndpointChainNameBase(svcPortName, strings.ToLower(string(svcPortName.Protocol)), baseInfo.Endpoint)
+	return &endpointInfo{
+		BaseEndpointInfo: baseInfo,
+		chainName:        servicePortEndpointChainNamePrefix + chainNameBase,
+		affinitySetName:  servicePortEndpointAffinityNamePrefix + chainNameBase,
+	}
+}
+
+// Proxier is an nftables based proxy
+type Proxier struct {
+	// ipFamily defines the IP family which this proxier is tracking.
+	ipFamily v1.IPFamily
+
+	// endpointsChanges and serviceChanges contains all changes to endpoints and
+	// services that happened since nftables was synced. For a single object,
+	// changes are accumulated, i.e. previous is state from before all of them,
+	// current is state after applying all of those.
+	endpointsChanges *proxy.EndpointsChangeTracker
+	serviceChanges   *proxy.ServiceChangeTracker
+
+	mu           sync.Mutex // protects the following fields
+	svcPortMap   proxy.ServicePortMap
+	endpointsMap proxy.EndpointsMap
+	nodeLabels   map[string]string
+	// endpointSlicesSynced, and servicesSynced are set to true
+	// when corresponding objects are synced after startup. This is used to avoid
+	// updating nftables with some partial data after kube-proxy restart.
+	endpointSlicesSynced bool
+	servicesSynced       bool
+	initialized          int32
```

Comment:
this was a nice addition to the std library https://pkg.go.dev/sync/atomic#Bool, the atomic.bool, it seems you can also remove some mutex with the *synced variables

Reply:
This is just copied from the iptables backend. I do have plans to clean this up eventually: https://github.com/kubernetes/kubernetes/pull/117436/commits/f0e5a48f25eca6506042e0fb228409e349644f86

Reply:
great

Reply:
Yes, I think this is another situation where there was initially a single variable that needed to be handled atomically, and then a few more got added and the code as a whole never got reconsidered.

If you look at the commit I linked to above though, my eventual plan is to just get rid of all of the initialization-time synchronization in the Proxiers and just have the Proxier methods not even get called until after the informers are fully synced.